File: 03.bools.hs

data Term = 
       TTrue
     | TFalse
     | IfThenElse Term Term Term

isval :: a -> Term -> Bool
isval _ TTrue = True
isval _ TFalse = True
isval _ _ = False

eval1 ctx (IfThenElse TTrue t2 t3) = do
      return t2

eval1 ctx (IfThenElse TFalse t2 t3) = do
      return t3

eval1 ctx (IfThenElse t1 t2 t3) = do
      t1' <- eval1 ctx t1; 
      return (IfThenElse t1' t2 t3)

eval1 _ _ = Nothing


eval ctx t = let t' = eval1 ctx t
             in case t' of
                Nothing -> t
                Just t'' -> eval ctx t''

File: 03.bools.pl

% true and false are builtin in prolog
ifthenelse(X,Y,Z) :- term(X), term(Y), term(Z).

term(true).
term(false).
term(ifthenelse(_,_,_)).

value(true).
value(false).

ev(ifthenelse(true, A, _), A) :- !.
ev(ifthenelse(false, _, B), B) :- !.
ev(ifthenelse(A, B, C), T) :- 
		ev(A, A2),
		ev(ifthenelse(A2, B, C), T).
